src/qute/languageServer/requirements.ts

```diff
@@ -33,48 +33,49 @@ export async function resolveRequirements(api: JavaExtensionAPI): Promise<Requir
 
     const javaHome = await checkJavaRuntime();
     const javaVersion = await checkJavaVersion(javaHome);
-    return Promise.resolve({tooling_jre: javaHome, tooling_jre_version: javaVersion, java_home: javaHome, java_version: javaVersion});
+    return Promise.resolve({ tooling_jre: javaHome, tooling_jre_version: javaVersion, java_home: javaHome, java_version: javaVersion });
 }
 
-function checkJavaRuntime(): Promise<string> {
-    return new Promise((resolve, reject) => {
-        let source: string;
-        let javaHome: string|undefined = readJavaHomeConfig();
+async function checkJavaRuntime(): Promise<string> {
+    let source: string;
+    let javaHome: string | undefined = await readJavaHomeConfig();
 
-        if (javaHome) {
-            source = 'The java.home variable defined in VS Code settings';
-        } else {
-            javaHome = process.env['JDK_HOME'];
-            if (javaHome) {
-                source = 'The JDK_HOME environment variable';
-            } else {
-                javaHome = process.env['JAVA_HOME'];
-                source = 'The JAVA_HOME environment variable';
-            }
+    if (javaHome) {
+        source = 'The java.home variable defined in VS Code settings';
+        javaHome = expandHomeDir(javaHome);
+        if (!fs.existsSync(javaHome)) {
+            throw openJDKDownload(source + ' points to a missing folder');
+        } else if (!fs.existsSync(path.resolve(javaHome as string, 'bin', JAVA_FILENAME))) {
+            throw openJDKDownload(source + ' does not point to a Java runtime.');
         }
+        return javaHome;
+    }
+    // No settings, let's try to detect as last resort.
+    const javaRuntimes = await findRuntimes({ withVersion: true, withTags: true });
+    if (javaRuntimes.length) {
+        sortJdksBySource(javaRuntimes);
+        javaHome = javaRuntimes[0].homedir;
+    } else {
+        openJDKDownload("Java runtime could not be located. Please download and install Java or use the binary server.");
```

Comment:
`throw` the result of this function, similar to how it's done earlier in the function